client/java/src/main/java/com/linecorp/centraldogma/client/MappingWatcher.java

```diff
@@ -0,0 +1,182 @@
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Maps;
+
+import com.linecorp.centraldogma.common.Revision;
+
+final class MappingWatcher<T, U> implements Watcher<U> {
+
+    private static final Logger logger = LoggerFactory.getLogger(MappingWatcher.class);
+
+    static <T, U> MappingWatcher<T, U> of(Watcher<T> parent, Function<? super T, ? extends U> mapper,
+                                          Executor executor, boolean closeParentWhenClosing) {
+        requireNonNull(parent, "parent");
+        requireNonNull(mapper, "mapper");
+        requireNonNull(executor, "executor");
+        // TODO(minwoo): extract mapper function and combine it with the new mapper.
+        return new MappingWatcher<>(parent, mapper, executor, closeParentWhenClosing);
+    }
+
+    private final Watcher<T> parent;
+    private final Function<? super T, ? extends U> mapper;
+    private final Executor mapperExecutor;
+    private final boolean closeParentWhenClosing;
+    private final CompletableFuture<Latest<U>> initialValueFuture = new CompletableFuture<>();
+    private final List<Entry<BiConsumer<? super Revision, ? super U>, Executor>> updateListeners =
+            new CopyOnWriteArrayList<>();
+
+    @Nullable
+    private volatile Latest<U> mappedLatest;
+    private volatile boolean closed;
+
+    MappingWatcher(Watcher<T> parent, Function<? super T, ? extends U> mapper, Executor mapperExecutor,
+                   boolean closeParentWhenClosing) {
+        this.parent = parent;
+        this.mapper = mapper;
+        this.mapperExecutor = mapperExecutor;
+        this.closeParentWhenClosing = closeParentWhenClosing;
+        parent.initialValueFuture().exceptionally(cause -> {
+            initialValueFuture.completeExceptionally(cause);
+            return null;
+        });
+        parent.watch((revision, value) -> {
+            if (closed) {
+                return;
+            }
+            final U mappedValue = mapper.apply(value);
+            final Latest<U> oldLatest = mappedLatest;
+            if (oldLatest != null && oldLatest.value() == mappedValue) {
+                return;
+            }
+
+            // mappedValue can be nullable which is fine.
+            final Latest<U> newLatest = new Latest<>(revision, mappedValue);
+            mappedLatest = newLatest;
+            notifyListeners(newLatest);
+            if (!initialValueFuture.isDone()) {
+                initialValueFuture.complete(newLatest);
```

Comment:
> If upstream notify fails for any reason, I think it is possible that initialValueFuture isn't completed. (exception while mapping, executor rejection when notifying, etc..)

I think we can use `try catch` for those cases. I forgot to catch the exception from the `mapper` but I intentionally remove the `RejectedExecutionException` cases because we don't do that in other code and I think it's really rare and only happens when a server shuts down. (If the watcher is closed before the server shuts down, that would not happen.)
Should we also catch the `RejectedExecutionException`?

Reply:
>  I think it's really rare and only happens when a server shuts down

I think it's fine to leave unhandled 😅  the previous behavior was to just close the watcher which was unexpected for me anyways.

> If the watcher is closed before the server shuts down, that would not happen.)

I do think it can still happen in other cases since the executor can be user provided. i.e. if the user intentionally passes an executor with queueSize = 1, maxPoolSize = 1 and adds a bunch of watchers... or the executor could be filled due to a different code path.
Having said this, this example is probably unrealistic + we probably don't have to worry about it 😄

Reply:
> Having said this, this example is probably unrealistic + we probably don't have to worry about it

That's what I have thought. 😄 
It really makes code dirty and we don't do the same approach on our codebase including Armeria. Let's get back to this later if we have a problem with it. 😄